dot/sync/chain_processor.go

```diff
@@ -95,99 +95,109 @@ func (s *chainProcessor) processReadyBlocks() {
-func (s *chainProcessor) processBlockData(bd *types.BlockData) error {
-	if bd == nil {
-		return ErrNilBlockData
-	}
+func (c *chainProcessor) processBlockData(blockData types.BlockData) error { //nolint:revive
+	logger.Debugf("processing block data with hash %s", blockData.Hash)
 
-	hasHeader, err := s.blockState.HasHeader(bd.Hash)
+	hasHeader, err := c.blockState.HasHeader(blockData.Hash)
 	if err != nil {
-		return fmt.Errorf("failed to check if block state has header for hash %s: %w", bd.Hash, err)
+		return fmt.Errorf("checking if block state has header: %w", err)
 	}
-	hasBody, err := s.blockState.HasBlockBody(bd.Hash)
+
+	hasBody, err := c.blockState.HasBlockBody(blockData.Hash)
 	if err != nil {
-		return fmt.Errorf("failed to check block state has body for hash %s: %w", bd.Hash, err)
+		return fmt.Errorf("checking if block state has body: %w", err)
 	}
 
 	if hasHeader && hasBody {
-		// TODO: fix this; sometimes when the node shuts down the "best block" isn't stored properly,
-		// so when the node restarts it has blocks higher than what it thinks is the best, causing it not to sync
-		// if we update the node to only store finalised blocks in the database, this should be fixed and the entire
-		// code block can be removed (#1784)
-		block, err := s.blockState.GetBlockByHash(bd.Hash)
+		err = c.processBlockDataWithStateHeaderAndBody(blockData)
 		if err != nil {
-			logger.Debugf("failed to get block header for hash %s: %s", bd.Hash, err)
-			return err
-		}
-
-		logger.Debugf(
-			"skipping block number %d with hash %s, already have",
-			block.Header.Number, bd.Hash) // TODO is this valid?
-
-		err = s.blockState.AddBlockToBlockTree(block)
-		if errors.Is(err, blocktree.ErrBlockExists) {
-			return nil
-		} else if err != nil {
-			logger.Warnf("failed to add block with hash %s to blocktree: %s", bd.Hash, err)
-			return err
+			return fmt.Errorf("processing block data with header and "+
+				"body in block state: %w", err)
 		}
+		return nil
+	}
 
-		if bd.Justification != nil {
-			logger.Debugf("handling Justification for block number %d with hash %s...", block.Header.Number, bd.Hash)
-			err = s.handleJustification(&block.Header, *bd.Justification)
-			if err != nil {
-				return fmt.Errorf("handling justification: %w", err)
-			}
+	if blockData.Header != nil && blockData.Body != nil {
+		err = c.processBlockDataWithHeaderAndBody(blockData)
+		if err != nil {
+			return fmt.Errorf("processing block data with header and body: %w", err)
 		}
+		logger.Debugf("block with hash %s processed", blockData.Hash)
+	}
 
-		// TODO: this is probably unnecessary, since the state is already in the database
-		// however, this case shouldn't be hit often, since it's only hit if the node state
-		// is rewinded or if the node shuts down unexpectedly (#1784)
-		state, err := s.storageState.TrieState(&block.Header.StateRoot)
+	if blockData.Justification != nil && len(*blockData.Justification) > 0 && blockData.Header != nil {
```

Comment:
supermeganit: in case header is nil will be not needed to derefer justification and check its len
```suggestion
	if blockData.Header != nil && blockData.Justification != nil && len(*blockData.Justification) > 0 {
```

Reply:
I did some benchmarks and this does not make it faster 😢

Reply:
I can confirm, I did this benchmark mixing half with header and half with no header and justification set.

```go
func BenchmarkXxx(b *testing.B) {
	withHeader := types.BlockData{
		Header:        &types.Header{},
		Justification: &[]byte{1, 2, 3, 4, 5},
	}
	withNoHeader := types.BlockData{
		Justification: &[]byte{1, 2, 3, 4, 5},
	}

	mixedBlockData := make([]types.BlockData, 500)
	for i := range mixedBlockData {
		if i%2 == 0 {
			mixedBlockData[i] = withHeader
		} else {
			mixedBlockData[i] = withNoHeader
		}
	}

	b.ResetTimer()

	b.Run("original", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			for _, blockData := range mixedBlockData {
				_ = blockData.Justification != nil && len(*blockData.Justification) > 0 && blockData.Header != nil
			}
		}
	})

	b.Run("suggested", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			for _, blockData := range mixedBlockData {
				_ = blockData.Header != nil && blockData.Justification != nil && len(*blockData.Justification) > 0
			}
		}
	})
}
```

I then even went the assembly route to check if there was any difference, and there is:

![image](https://user-images.githubusercontent.com/15654980/195523945-3e324105-96a1-4283-9fa3-719a6fe0af95.png)

So at least that's that :wink: I guess our CPUs are just too quick.